gordon/metrics/ffwd.py

```diff
@@ -0,0 +1,138 @@
+
+from gordon import interfaces
+
+
+class FfwdClientProtocol(asyncio.DatagramProtocol):
+    """Protocol for sending one-off messages to ffwd.
+
+    Args:
+        message (bytes): Message for ffwd agent.
+    """
+    def __init__(self, message):
+        self.message = message
+        self.transport = None
+
+    def connection_made(self, transport):
+        self.transport = transport
+        self.transport.sendto(self.message)
+        self.transport.close()
+
+
+class UDPClient:
+    """Client for sending UDP datagrams.
+
+    Args:
+        ip (str): Destination IP address.
+        port (int): Destination port.
+        loop (asyncio.AbstractEventLoop impl): (optional) Event loop.
+    """
+    def __init__(self, ip, port, loop=None):
+        self.ip = ip
+        self.port = port
+        self.loop = loop or asyncio.get_event_loop()
+
+    async def send(self, metric):
+        """Transform metric to JSON bytestring and send to server.
+
+        Args:
+            metric (dict): Complete metric to send as JSON.
+        """
+        message = json.dumps(metric).encode('utf-8')
+        await self.loop.create_datagram_endpoint(
+            lambda: FfwdClientProtocol(message),
+            remote_addr=(self.ip, self.port))
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class FfwdTimer:
+    """ITimer implementation which sends UDP messages to FFWD on completion.
+
+    Args:
+        metric (dict): Dict representation of the metric to send.
+        udp_client (UDPClient): A metric sending client.
+        time_unit (number): (optional) Scale time unit for use with time.time()
+            (example: 1E9 to send nanoseconds).
+    """
+    def __init__(self, metric, udp_client, time_unit=1):
+        self.metric = metric
+        self.udp_client = udp_client
+        self.time_unit = time_unit
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    async def start(self):
+        self._start_time = time.time()
+
+    async def stop(self):
+        time_elapsed = (time.time() - self._start_time) * self.time_unit
+        self.metric['value'] = time_elapsed
+        await self.udp_client.send(self.metric)
+
+
+@zope.interface.implementer(interfaces.IMetricRelay)
+class FfwdRelay:
+    """IMetricRelay implementation which sends metrics to FFWD immediately.
+
+    Args:
+        config (dict): Required keys:
+            key (str): A key to identify the service as a whole.
+            ffwd_ip (str): ffwd IP address.
+            ffwd_port (int): ffwd port number.
+            time_unit (number): Scale unit to use with timers.
+    """
+    def __init__(self, config):
+        self.key = config['key']
+        self.udp_client = UDPClient(config['ffwd_ip'], config['ffwd_port'])
+        self.time_unit = config['time_unit']
+
+    def _create_metric(self, metric_name, value, **kwargs):
+        attrs = kwargs.get('attrs') or {}
```

Comment:
If someone passes is explicitly passes`attrs=None`, I want to be sure I don't put None in the json.  I actually don't know for sure that ffwd would freak out. I can look into it.

Reply:
Eh, fair enough.

Reply:
Hmm maybe I'm under-caffeinated, so please correct me if I'm wrong: as I'm instrumenting gordon core, I'm doing something like this:

```py
attrs = {'error': 'invalid-msg-provider'}
self.metrics.incr('dropped-message', attrs=attrs)
```

The `attrs` kwarg is specific to ffwd/our use case, so it seems like it's a bit leaky, i.e. if we switch for prometheus implementation, `attrs` will be ignored (hopefully not throw an error), and therefore losing valuable context.

I don't know how to handle tags (actually, I don't know if we really use them?), but perhaps we can assume all kwargs are attributes?

And maybe it'd be more clear if we renamed `kwargs` to `context` or `extra_context`?

Reply:
> if we switch for prometheus implementation, attrs will be ignored

I think prometheus wouldn't ignore attrs since it's similar to the concept of [labels][0] - key:values. We could rename it something more generic, but clearer (_attributes_?) to better fit how different dbs call it (heroic - attributes, prometheus - labels, influx - fields) and maybe make it a required argument?

[0]: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

Reply:
Yeah, I was going for non-breaking, but that data loss was acceptable.  I wasn't sure if all other libraries supported additional data, so I thought data loss might be inherent in that case.  

I'm totally for putting attributes back on the Interface def though (named attributes or context or whatever) or doing the kwargs == attrs (even though that doesn't guarantee any implementer will actually use them).  In the kwargs case we would lose tags, but as you mentioned, we're not using them anywhere anyways. We could just drop them for all cases.  

Is the `context` option, the one everybody can agree on @econchick @matt0bi? It works for me.

Reply:
:+1: for the `context` option @descent3me.

Reply:
Just to be clear, which do we like?

1) 
```
def timer(metric_name, context):
```
2)
```
def timer(metric_name, context=None):
```
3) 
```
def timer(metric_name, context=None, **kwargs):
``` 
4) 
```
def timer(metric_name, context, **kwargs):
``` 
5) 
```
def timer(metric_name, **context):
```

I was imagining 4) and I want to make sure we're on the same page.

Reply:
I was thinking of either 3 or 4, I think either one would work alright.

Reply:
I was thinkin 3 in case an implementation wants/needs to completely ignore it.

Reply:
Works for me! Also, I just noticed none of these have any args descriptions, so I'll add that indicating that if used, context should be a dict.